File: src/main/scala/com/raphtory/algorithms/newer/ConnectedComponents.scala
package com.raphtory.algorithms.newer

import com.raphtory.core.model.algorithm.{GraphAlgorithm, GraphPerspective, Row}

class ConnectedComponents(path:String) extends GraphAlgorithm{
  override def algorithm(graph: GraphPerspective): Unit = {
    val nodeCount = graph.nodeCount()
    graph
      .step({
        vertex =>
          vertex.setState("cclabel", vertex.ID)
          vertex.messageAllNeighbours(vertex.ID)
      })
      .iterate({
        vertex =>
          val label = vertex.messageQueue[Long].min
          if (label < vertex.getState[Long]("cclabel")) {
            vertex.setState("cclabel", label)
            vertex messageAllNeighbours label
          }
          else
            vertex.voteToHalt()
      }, 100)
      .select(vertex => Row(vertex.ID(),vertex.getState[Long]("cclabel")))
      //.filter(r=> r.get(0).asInstanceOf[Long]==18174)
      .writeTo(path)
  }
}
object ConnectedComponents{
  def apply(path:String) = new ConnectedComponents(path)
}

File: python/src/pyraphtory/algorithms/connectedcomponents.py
from pyraphtory.algorithm import PyAlgorithm
from pyraphtory.graph import TemporalGraph, Row, Table, KeyPair
from pyraphtory.vertex import Vertex

CC_LABEL = 'cclabel'


class ConnectedComponents(PyAlgorithm):
    def __call__(self, graph: TemporalGraph) -> TemporalGraph:
        def step(v: Vertex):
            v[CC_LABEL] = v.id()
            v.message_all_neighbours(v.id())

        def iterate(v: Vertex):
            label = min(v.message_queue())
            if label < v[CC_LABEL]:
                v[CC_LABEL] = label
                v.message_all_neighbours(label)
            else:
                v.vote_to_halt()

        return graph.step(step).iterate(iterate, 100, True)

    def tabularise(self, graph: TemporalGraph) -> Table:
        return graph.select(lambda v: Row(KeyPair("name",v.name()), KeyPair(CC_LABEL, v[CC_LABEL])))
